### PR TITLE
fix containerized function mounts issue

### PR DESCRIPTION
**Pull request by natasha41575**
Fix https://github.com/kubernetes-sigs/kustomize/issues/4290

Kustomize was not mounting files into containerized functions directly because the `StorageMounts` field was not being properly propagated to the function filter. 

This PR allows users to use the `StorageMounts` fields, with the following requirements:
- mount paths cannot be absolute paths
- mount paths must be relative to the kustomization that refers to them
- mount paths must be under the kustomization directory that refers to them

As a side note, if we are going to be promoting functions in kustomize we should probably build up a strong test suite for containerized functions and run those tests in CI.  All tests needing docker are currently being skipped.

/cc @yuwenma 
/cc @KnVerey

### CODE REVIEW

File: api/go.mod [hunk 1]
```diff
@@ -14,3 +14,5 @@ require (
 	sigs.k8s.io/kustomize/kyaml v0.13.3
 	sigs.k8s.io/yaml v1.2.0
 )
+
+replace sigs.k8s.io/kustomize/kyaml => ../kyaml
```
Comment (yuwenma) [inline api/go.mod:18]: clean up?

File: api/krusty/fnplugin_test.go [hunk 2]
```diff
@@ -629,3 +631,148 @@ metadata:
...
+      releaseName: test
+      valuesFile: /tmp/charts/helloworld-values/values.yaml
+`)))
+	m, err := b.Run(
+		fSys,
+		tmpDir.String())
+	if utils.IsErrTimeout(err) {
+		// Don't fail on timeouts.
+		t.SkipNow()
+	}
```
Comment (yuwenma) [inline api/krusty/fnplugin_test.go:675]: Suggest not doing this on test since it can either introduce flakiness or bypass covering for corner cases. Do you know which step may cause the timeout? would it be possible to mock it?

File: api/krusty/testdata/charts/helloworld-chart/Chart.yaml [hunk 3]
```diff
@@ -0,0 +1,21 @@
+apiVersion: v2
```
Comment (yuwenma) [inline api/krusty/testdata/charts/helloworld-chart/Chart.yaml:1]: qq: shall we add headers for test files?

File: api/krusty/testdata/charts/helloworld-chart/values.yaml [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+  # We usually recommend not to specify default resources and to leave this as a conscious
+  # choice for the user. This also increases chances charts run on environments with little
+  # resources, such as Minikube. If you do want to specify resources, uncomment the following
+  # lines, adjust them as necessary, and remove the curly braces after 'resources:'.
+  # limits:
+  #   cpu: 100m
+  #   memory: 128Mi
+  # requests:
+  #   cpu: 100m
+  #   memory: 128Mi
```
Comment (yuwenma) [inline api/krusty/testdata/charts/helloworld-chart/values.yaml:62]: clean up? seems not relevant. Here and other places.

**Comment by yuwenma** [comment]
Some minor comments. Looks great!

File: api/krusty/testdata/charts/helloworld-chart/Chart.yaml (hunk 3 above)
Comment (natasha41575) [reply api/krusty/testdata/charts/helloworld-chart/Chart.yaml:1]: This is not my helm chart, I took it from kpt functions repo... I will have to check where it came from.

Comment (yuwenma) [reply api/krusty/testdata/charts/helloworld-chart/Chart.yaml:1]: oh, nvm then :)

File: api/go.mod (hunk 1 above)
Comment (natasha41575) [reply api/go.mod:18]: This is necessary while in development mode. We usually make this change after every release, I must have forgotten to do it after the last release

Comment (yuwenma) [reply api/go.mod:18]: oh oh I see. Basically the api and other kyaml/libs (kio, fn) are always tied but released separately.

File: api/krusty/testdata/charts/helloworld-chart/values.yaml (hunk 4 above)
Comment (natasha41575) [reply api/krusty/testdata/charts/helloworld-chart/values.yaml:62]: done!

File: api/krusty/fnplugin_test.go (hunk 2 above)
Comment (natasha41575) [reply api/krusty/fnplugin_test.go:675]: Removed! I don't think it's necessary. Mistakenly copied the lines from another test, thanks for the catch.

File: api/krusty/fnplugin_test.go [hunk 5]
```diff
@@ -629,3 +630,144 @@ metadata:
...
+generators:
+  - |-
+    apiVersion: v1alpha1
+    kind: RenderHelmChart
+    metadata:
+      name: demo
+      annotations:
+        config.kubernetes.io/function: |
+          container:
+            image: gcr.io/kpt-fn/render-helm-chart:v0.1.0
```
Comment (KnVerey) [inline api/krusty/fnplugin_test.go:658]: This makes our CI depend on this Kpt container continuing to be published here, right? Can you open an issue about switching to the donated one once it is available?

Comment (natasha41575) [reply api/krusty/fnplugin_test.go:658]: > This makes our CI depend on this Kpt container continuing to be published here, right? Can you open an issue about switching to the donated one once it is available?

This was on my mind too, I can file an issue to make sure we don't forget. Though this is not a problem until we start to run docker tests in our CI. I'll file an issue for that as well.

**Comment by KnVerey** [comment]
I poked around the history of the problem file, and it seems to me this _never_ worked. Is that correct?

If so, if we were to release this PR, it would introduce a new security issue that (intended or not) we didn't have before. It looks like the feature was originally intended to have restrictions on the path, but that did not get implemented: https://github.com/kubernetes-sigs/kustomize/issues/1902

Since container functions ARE enabled in kubectl, we should be very careful about this decision.

> volumes are relative to the Resource config
> volumes must be in a directory under the config directory (no ../)

The way Plugin Graduation proposes to handle this is through Catalog, where the Catalog entry must specify allowed mounts. Inline config like what we're turning on here actually wouldn't be supported at all anymore:

> If a containerized plugin needs network or disk access, its catalog entry MUST specify it. Redundant inline provider config cannot be used to grant it. When the catalog does specify it, no additional flags or config fields will be required to grant it the required access.

**Comment by natasha41575** [comment]
> I poked around the history of the problem file, and it seems to me this never worked. Is that correct?

It seems that this is the case, and I agree that you're right, it probably isn't great to allow an unrestricted path to be mounted in the container. 

However, this lack of mounts would block usage of the render-helm-chart KRM function for local charts and I don't think it would be practical to wait for the Plugin Graduation and Catalog to be completed before asking users to migrate from the builtin Helm plugin. 

WDYT about restricting mounted paths to the same load restrictions as all other paths (i.e. it must be in the current kustomization directory)? If we do that, would that alleviate your concerns regarding security?

**Comment by KnVerey** [comment]
> WDYT about restricting mounted paths to the same load restrictions as all other paths (i.e. it must be in the current kustomization directory)? If we do that, would that alleviate your concerns regarding security?

Yes, this would work for me.

**Comment by natasha41575** [comment]
@KnVerey awesome, PR is updated, with a check to ensure that the mounted paths are in the current directory. I added a test `TestFnContainerMountsLoadRestrictions` for this restriction as well.

**Comment by KnVerey** [comment]
> As a side note, if we are going to be promoting functions in kustomize we should probably build up a strong test suite for containerized functions and run those tests in CI. All tests needing docker are currently being skipped.

As part of #4564 I noticed that it seems like _some_ Docker-based tests do run in CI. But only the ones in cmd/config, only if they're gated with `KUSTOMIZE_DOCKER_E2E` specifically, and only on Linux: https://github.com/kubernetes-sigs/kustomize/blob/master/.github/workflows/go.yml#L61. Since we aren't running in verbose mode right now you can't see it in this PR's tests, but check out this test run from my PR: https://github.com/kubernetes-sigs/kustomize/runs/5826226816?check_suite_focus=true#step:4:846

File: kyaml/runfn/runfn.go [hunk 6]
```diff
@@ -468,11 +468,26 @@ func (r *RunFns) ffp(spec runtimeutil.FunctionSpec, api *yaml.RNode, currentUser
...
 		}
+
+		// Storage mounts can either come from kustomize fn run --mounts,
+		// or from the declarative function mounts field.
+		// In the latter case, we ensure that the mount path is subject
+		// to kustomize's load restrictions.
+		storageMounts := spec.Container.StorageMounts
+		for _, mount := range storageMounts {
+			if strings.HasPrefix(filepath.Clean(mount.Src), "../") {
+				return nil, fmt.Errorf("mount path %s is not permitted; declarative mount paths must be under the current directory", mount.Src) //nolint:goerr113
```
Comment (KnVerey) [inline kyaml/runfn/runfn.go:479]: What's the linter failure here? Since we're in kyaml, should we be using the errors package instead of fmt anyway?

File: kyaml/runfn/runfn.go [hunk 7]
```diff
@@ -468,11 +468,26 @@ func (r *RunFns) ffp(spec runtimeutil.FunctionSpec, api *yaml.RNode, currentUser
...
+			}
+		}
+		if len(storageMounts) == 0 {
```
Comment (KnVerey) [inline kyaml/runfn/runfn.go:482]: So declarative always overrides imperative? I think I'd expect it to be the other way around, or for imperative to be appended. I'm guessing we don't document one way or the other, currently.

Comment (natasha41575) [reply kyaml/runfn/runfn.go:482]: I'll change the logic to append it. That sounds like more intuitive behavior to me.

File: kyaml/runfn/runfn.go (hunk 6 above)
Comment (natasha41575) [reply kyaml/runfn/runfn.go:479]: Ah yeah you're right, I'll change it to use `errors`, that should make the linter failure go away. Thanks!

**Comment by natasha41575** [comment]
> As part of #4564 I noticed that it seems like some Docker-based tests do run in CI. But only the ones in cmd/config, only if they're gated with KUSTOMIZE_DOCKER_E2E specifically, and only on Linux: https://github.com/kubernetes-sigs/kustomize/blob/master/.github/workflows/go.yml#L61. Since we aren't running in verbose mode right now you can't see it in this PR's tests, but check out this test run from my PR: https://github.com/kubernetes-sigs/kustomize/runs/5826226816?check_suite_focus=true#step:4:846

Will your rework of the CI fix that? I'm wondering if docker is available in the Github Actions environment, in which case I think all the tests gated by [skipIfNoDocker](https://github.com/kubernetes-sigs/kustomize/blob/49c0ed13263ed75e2a7c5d444479fda85f29d993/api/krusty/fnplugin_test.go#L216) will run.

**Comment by natasha41575** [comment]
The CI failure here should be fixed by  https://github.com/kubernetes-sigs/kustomize/pull/4564. I think it's because `go get` needs to be changed to `go install`.

**Comment by KnVerey** [comment]
> Will your rework of the CI fix that? I'm wondering if docker is available in the Github Actions environment, in which case I think all the tests gated by [skipIfNoDocker](https://github.com/kubernetes-sigs/kustomize/blob/49c0ed13263ed75e2a7c5d444479fda85f29d993/api/krusty/fnplugin_test.go#L216) will run.

Oh, interesting. My PR isn't specifically trying to do anything with the docker tests, but I think you're right that it inadvertently will run more of them, just by virtue of shifting all non-plugin tests to the GHA env. WDYT about merging mine first, rebasing this one, and confirming by looking for your new tests in the logs? Then as a follow up perhaps we can switch _all_ the gating to use skipIfNoDocker instead of KUSTOMIZE_DOCKER_E2E

File: kyaml/runfn/runfn.go [hunk 8]
```diff
@@ -468,11 +468,24 @@ func (r *RunFns) ffp(spec runtimeutil.FunctionSpec, api *yaml.RNode, currentUser
...
 		}
+
+		// Storage mounts can either come from kustomize fn run --mounts,
+		// or from the declarative function mounts field.
+		// In the latter case, we ensure that the mount path is subject
+		// to kustomize's load restrictions.
+		storageMounts := spec.Container.StorageMounts
+		for _, mount := range storageMounts {
+			if strings.HasPrefix(filepath.Clean(mount.Src), "../") {
+				return nil, errors.Errorf("mount path %s is not permitted; declarative mount paths must be under the current directory", mount.Src)
```
Comment (KnVerey) [inline kyaml/runfn/runfn.go:479]: I was just sanity checking against the Starlark file version of this constraint, and it handles another case: absolute paths. If we don't ban those, we'll need to be comparing more specifically to the kustomize root.

Comment (natasha41575) [reply kyaml/runfn/runfn.go:479]: Dug a little bit and I think we have to allow absolute paths. With a relative path, I get this error from docker:

`docker: Error response from daemon: invalid mount config for type "bind": invalid mount path: 'testdata/charts' mount path must be absolute.`

I'll update this check here to test for absolute paths.

Comment (natasha41575) [reply kyaml/runfn/runfn.go]: Updated!

**Comment by natasha41575** [comment]
> WDYT about merging mine first, rebasing this one, and confirming by looking for your new tests in the logs? Then as a follow up perhaps we can switch all the gating to use skipIfNoDocker instead of KUSTOMIZE_DOCKER_E2E

SGTM.

/hold

File: api/krusty/fnplugin_test.go [hunk 9]
```diff
@@ -630,3 +631,186 @@ metadata:
...
+	assert.NoError(t, err)
+
+	path, err := os.Getwd()
+	assert.NoError(t, err)
+
+	src := filepath.Join(path, "testdata", "charts") + string(filepath.Separator)
+	dst := filepath.Join(tmpDir.String(), "charts") + string(filepath.Separator)
+	assert.NoError(t, os.Mkdir(dst, 0777))
+
+	cmd := exec.Command("cp", "-r", src, dst)
```
Comment (natasha41575) [inline api/krusty/fnplugin_test.go:652]: This test passes locally, but not in the Linux CI job. I'll dig into this again tomorrow.

File: api/krusty/fnplugin_test.go [hunk 10]
```diff
@@ -630,3 +632,194 @@ metadata:
...
+	assert.NoError(t, os.MkdirAll(dst, 0777))
+
+	if runtime.GOOS == "darwin" {
+		cmd := exec.Command("cp", "-r", src, dst)
+		assert.NoError(t, cmd.Run())
+	} else if runtime.GOOS == "linux" {
+		d := dst + "."
+		cmd := exec.Command("cp", "-r", src, d)
+		assert.NoError(t, cmd.Run())
+	} else {
```
Comment (natasha41575) [inline api/krusty/fnplugin_test.go:660]: This test won't get run on Windows. But since the krusty tests don't run in the Windows CI job, and neither of us have a Windows machine, I don't have any method of actually verifying that whatever I write here would work on windows. I'm thinking we can reenable this once we're able to run Windows tests somewhere.

File: kyaml/runfn/runfn.go [hunk 11]
```diff
@@ -313,6 +313,14 @@ func (r RunFns) getFunctionFilters(global bool, fns ...*yaml.RNode) (
 			// resource doesn't have function spec
 			continue
 		}
+		// convert declarative relative paths to absolute (otherwise docker will throw an error)
```
Comment (natasha41575) [inline kyaml/runfn/runfn.go:316]: Some justification for doing this:

- it makes sense to allow our declarative mount paths in kustomize to be relative paths because we treat the kustomization as a single portable configuration that in theory should be able to be moved to different directories
- the reasons that Docker doesn't allow bind mount paths to be relative are [described here](https://github.com/docker/cli/issues/1203#issuecomment-981757086), and it seems mostly like this is because bind mounts are performed on the daemon side. I don't think that means that we should disallow relative paths on the kustomize side if conversion is easy for us to do.

File: api/internal/plugins/loader/loader.go [hunk 12]
```diff
@@ -225,9 +225,26 @@ func (l *Loader) makeBuiltinPlugin(r resid.Gvk) (resmap.Configurable, error) {
...
+			var err error
+			if filepath.IsAbs(mount.Src) {
+				relPath, err = filepath.Rel(ldr.Root(), mount.Src)
+				if err != nil {
+					return nil, errors.Wrap(err, "could not convert mount path to relative path")
+				}
+			} else {
+				relPath = mount.Src
+			}
+			if strings.HasPrefix(filepath.Clean(relPath), "../") {
```
Comment (KnVerey) [inline api/internal/plugins/loader/loader.go:243]: I agree with your other comment that relative paths should be not only allowed but encouraged, since we want Kustomizations to be encapsulated. Doesn't this still prevent that?

File: api/internal/plugins/loader/loader.go [hunk 13]
```diff
@@ -225,9 +225,26 @@ func (l *Loader) makeBuiltinPlugin(r resid.Gvk) (resmap.Configurable, error) {
...
 
-func (l *Loader) loadPlugin(res *resource.Resource) (resmap.Configurable, error) {
+func (l *Loader) loadPlugin(ldr ifc.Loader, res *resource.Resource) (resmap.Configurable, error) {
 	spec := fnplugin.GetFunctionSpec(res)
 	if spec != nil {
+		// validation check that function mounts are under the current kustomization directory
+		for _, mount := range spec.Container.StorageMounts {
+			var relPath string
+			var err error
+			if filepath.IsAbs(mount.Src) {
```
Comment (KnVerey) [inline api/internal/plugins/loader/loader.go:235]: WDYT about disallowing absolute paths entirely, on the grounds of them breaking encapsulation, as you pointed out?

File: kyaml/runfn/runfn.go (hunk 11 above)
Comment (KnVerey) [reply kyaml/runfn/runfn.go:316]: I agree with allowing paths to be relative (in fact I'm thinking we should require them to be), and I understand the need to convert them to absolute for Docker. But don't we need to do that at some point where we have the Kustomization root? Or else we're expanding them relative to the current working directory, I assume. I think we should treat relative paths here consistently with what we decided for exec paths.

File: api/krusty/fnplugin_test.go [hunk 14]
```diff
@@ -630,3 +632,195 @@ metadata:
...
+	dst := filepath.Join(tmpDir.String(), "testdata", "charts") + string(filepath.Separator)
+	assert.NoError(t, os.MkdirAll(dst, 0777))
+
+	switch runtime.GOOS {
+	case "darwin":
+		cmd := exec.Command("cp", "-r", src, dst)
+		assert.NoError(t, cmd.Run())
+	case "linux":
+		d := dst + "."
+		cmd := exec.Command("cp", "-r", src, d)
```
Comment (natasha41575) [inline api/krusty/fnplugin_test.go:659]: FYI this was failing in CI, which means the CI job ran it!

File: kyaml/runfn/runfn.go (hunk 11 above)
Comment (natasha41575) [reply kyaml/runfn/runfn.go]: > But don't we need to do that at some point where we have the Kustomization root?

Good catch! I've updated the PR so that the working directory gets propagated correctly down to the containerized function filter. I've done manual testing to make sure that the path is resolved against the kustomization root.

File: api/internal/plugins/loader/loader.go (hunk 12 above)
Comment (natasha41575) [reply api/internal/plugins/loader/loader.go]: > Doesn't this still prevent that?

Could you elaborate a bit more? I'm not sure that I understand what you mean, this particular line will throw an error of the relative path points to something outside of the current directory. 

I did modify this code to disallow absolute paths if that addresses your concerns.

File: api/internal/plugins/loader/loader.go (hunk 13 above)
Comment (natasha41575) [reply api/internal/plugins/loader/loader.go:233]: done!

File: api/internal/plugins/loader/loader.go [hunk 15]
```diff
@@ -49,7 +49,9 @@ func (l *Loader) Config() *types.PluginConfig {
 
 // SetWorkDir sets the working directory for this loader's plugins
 func (l *Loader) SetWorkDir(wd string) {
-	l.pc.FnpLoadingOptions.WorkingDir = wd
+	if wd != string(filepath.Separator) {
+		l.pc.FnpLoadingOptions.WorkingDir = wd
```
Comment (natasha41575) [inline api/internal/plugins/loader/loader.go:53]: Without this change, `TestFnContainerEnvVars` fails with the error `couldn't execute function: root working directory '/' not allowed`. `ldr.Root()` returns "/" when run at the root-level, i.e. whenever we do `th.Run(".")` in our tests. If we don't set the working dir here, then the function filter uses the current working directory, which I believe is correct in the cases that will hit this problem.

File: kyaml/fn/runtime/container/container_test.go [hunk 16]
```diff
@@ -68,7 +70,8 @@ metadata:
 		},
 
 		{
-			name: "storage_mounts",
+			name:          "storage_mounts",
+			skipOnWindows: true, // the fieldpaths are different on Windows
```
Comment (natasha41575) [inline kyaml/fn/runtime/container/container_test.go:74]: Alternatively, I could separate the tests two different files (one for Windows and one for everything else), and have different "expected" values for each

File: api/krusty/testdata/fnmounts/kustomization.yaml [hunk 17]
```diff
@@ -0,0 +1,20 @@
+# Test file for TestFnContainerMounts
```
Comment (natasha41575) [inline api/krusty/testdata/fnmounts/kustomization.yaml:1]: Trying to copy the charts directory into the temporary directory where the test harness was running was a bit of a headache for mac vs linux, so I instead created a test kustomization file here so that the test can run `kustomize build` directly on it.

**Comment by natasha41575** [comment]
According to the logs, the docker tests are running in the linux job but not the mac one. Filed https://github.com/kubernetes-sigs/kustomize/issues/4579

File: api/internal/plugins/loader/loader.go (hunk 15 above)
Comment (KnVerey) [reply api/internal/plugins/loader/loader.go:53]: Hmm, but isn't this just bypassing a real restriction we have in our function code? In other words, won't it really enable the root directory to be used as valid plugin load location, effectively allowing plugins from anywhere on the disk? 

> If we don't set the working dir here, then the function filter uses the current working directory, which I believe is correct behavior in the cases that hit this problem.

Can you explain more why you believe this is correct? My inclination would be that it's easier to understand what's happening if the behavior is always the same (kustomization root, not invocation dir).

> Alternatively I think we could rewrite TestFnContainerEnvVars so that it does th.Run("dir") or something, and then we won't have to have this change here.

Is it true that the tests in question are always using a virtual filesystem, and that's why it's not bad for the Kustomization dir to be the root? If so, that change will make the tests more realistic, right?

Comment (natasha41575) [reply api/internal/plugins/loader/loader.go:53]: 🤦‍♀️ you're right on all points. I think my brain was in another location when I wrote this part, for some reason I was thinking of the working directory as the kustomization root, which is obviously not the case and is the whole reason we are doing this anyways.

I will update the test.

Comment (natasha41575) [reply api/internal/plugins/loader/loader.go]: Test updated! I had to change it to use a temporary directory in the real file system because it's shelling out to docker and needs the working directory to be a real one.

File: api/krusty/fnplugin_test.go [hunk 18]
```diff
@@ -622,14 +628,198 @@ data:
...
+	_, err = b.Run(
+		fSys,
+		tmpDir.String())
+	assert.Error(t, err)
+	assert.Contains(t, err.Error(), "loading generator plugins: plugin RenderHelmChart."+
+		"v1alpha1.[noGrp]/demo.[noNs] with mount path '/tmp/dir' is not permitted; mount paths must"+
+		" be relative to the current kustomization directory")
+}
+
+func TestFnContainerMountsLoadRestrictions_outsideCurrentDir(t *testing.T) {
```
Comment (KnVerey) [inline api/krusty/fnplugin_test.go:794]: WDYT about adding another error case test for the root-as-Kustomization-dir exclusion specifically?

File: kyaml/runfn/runfn.go [hunk 19]
```diff
@@ -330,6 +330,7 @@ func (r RunFns) getFunctionFilters(global bool, fns ...*yaml.RNode) (
 		}
 		cf, ok := c.(*container.Filter)
 		if global && ok {
+			cf.Exec.WorkingDir = r.WorkingDir
```
Comment (KnVerey) [inline kyaml/runfn/runfn.go:333]: What does global mean here / is it right to only do this when it is true?

File: kyaml/fn/runtime/container/container_test.go (hunk 16 above)
Comment (natasha41575) [reply kyaml/fn/runtime/container/container_test.go]: It was more than the separator, I think the whole absolute path looks different. But using the filepath package is a good idea, I'm going to try using filepath.Join or potentially filepath.Abs if that doesn't work.

File: api/krusty/testdata/fnmounts/charts/helloworld-chart/values.yaml [hunk 20]
```diff
@@ -0,0 +1,27 @@
+replicaCount: 1
```
[the remark this answers is not in the document]
Comment (natasha41575) [reply api/krusty/testdata/fnmounts/charts/helloworld-chart/values.yaml]: Pared this down to a minimal example!

File: api/krusty/fnplugin_test.go [hunk 21]
```diff
@@ -622,14 +628,198 @@ data:
...
+}
+
+func TestFnContainerMounts(t *testing.T) {
+	skipIfNoDocker(t)
+
+	path, err := os.Getwd()
+	assert.NoError(t, err)
+
+	testDir := filepath.Join(path, "testdata", "fnmounts")
+	command := exec.Command("kustomize", "build", testDir, "--enable-alpha-plugins")
```
[the remark this answers is not in the document]
Comment (natasha41575) [reply api/krusty/fnplugin_test.go]: Originally I was trying to copy the entire testdata/fnmounts directory into the temp directory where the test framework was running, which turned out to be a headache.

But with a pared down helm chart (per your other suggestion), this is much easier to do with the test framework. Updated!

File: api/krusty/fnplugin_test.go (hunk 18 above)
Comment (natasha41575) [reply api/krusty/fnplugin_test.go:745]: added!

File: kyaml/runfn/runfn.go (hunk 19 above)
Comment (natasha41575) [reply kyaml/runfn/runfn.go]: Thanks for the catch! Not sure what global means, but I think it would make sense to always propagate down the working directory, so I moved it out.

File: kyaml/fn/runtime/container/container_test.go (hunk 16 above)
Comment (natasha41575) [reply kyaml/fn/runtime/container/container_test.go]: It worked! Thanks for the suggestion.
